File: LeetCode/146-lru-cache/LRUCache.java
import java.util.HashMap;

/**
 * A custom-built LRU (Least Recently Used) Cache using a
 * doubly linked list and a HashMap.
 */
class LRUCache {

  /**
   * Node class representing each entry in the LRU Cache.
   */
  public class Node {
    int key; // Key for the cache entry
    int val; // Value for the cache entry
    Node next; // Pointer to the next node in the linked list
    Node prev; // Pointer to the previous node in the linked list

    /**
     * Constructs a new Node with the specified key and value.
     * 
     * @param key The key for the cache entry.
     * @param val The value for the cache entry.
     */
    public Node(int key, int val) {
      this.key = key;
      this.val = val;
      this.next = null;
      this.prev = null;
    }
  }

  private int capacity;
  private HashMap<Integer, Node> map;
  private Node head;
  private Node tail;

  /**
   * Constructs an LRUCache with the specified capacity.
   * 
   * @param capacity The maximum number of items the cache can hold.
   */
  public LRUCache(int capacity) {
    this.capacity = capacity;
    this.map = new HashMap<>();
    this.head = new Node(0, 0);
    this.tail = new Node(0, 0);
    head.next = tail;
    tail.prev = head;
  }

  /**
   * Adds a node to the end of the doubly linked list.
   * 
   * @param node The node to be added.
   */
  public void add(Node node) {
    Node previousEnd = tail.prev;
    previousEnd.next = node;
    node.prev = previousEnd;
    node.next = tail;
    tail.prev = node;
  }

  /**
   * Removes a node from the doubly linked list.
   * 
   * @param node The node to be removed.
   */
  public void remove(Node node) {
    node.prev.next = node.next;
    node.next.prev = node.prev;
  }

  /**
   * Retrieves the value associated with the specified key.
   * 
   * @param key The key whose associated value is to be returned.
   * @return The value associated with the specified key, or -1 if the key does
   *         not exist.
   */
  public int get(int key) {
    if (!map.containsKey(key)) {
      return -1;
    }

    Node node = map.get(key);
    remove(node);
    add(node);
    return node.val;
  }

  /**
   * Adds a key-value pair to the cache.
   * 
   * @param key   The key with which the specified value is to be associated.
   * @param value The value to be associated with the specified key.
   */
  public void put(int key, int value) {
    if (map.containsKey(key)) {
      Node oldNode = map.get(key);
      remove(oldNode);
    }

    Node node = new Node(key, value);
    map.put(key, node);
    add(node);

    if (map.size() > capacity) {
      Node nodeToDelete = head.next;
      remove(nodeToDelete);
      map.remove(nodeToDelete.key);
    }
  }
}

File: LeetCode/146-lru-cache/LRUCache.ts
/**
 * Represents a node in the LRU cache.
 */
class CacheNode {
  key: number; // Key for the cache entry
  val: number; // Value for the cache entry
  next: CacheNode | null; // Pointer to the next node in the linked list
  prev: CacheNode | null; // Pointer to the previous node in the linked list

  /**
   * Constructs a new CacheNode with the specified key and value.
   * @param {number} key - The key for the cache entry.
   * @param {number} val - The value for the cache entry.
   */
  constructor(key: number, val: number) {
    this.key = key;
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

/**
 * LRU (Least Recently Used) Cache implemented with a doubly linked list and a Map.
 */
class LRUCache {
  private capacity: number;
  private map: Map<number, CacheNode>;
  private head: CacheNode;
  private tail: CacheNode;

  /**
   * Constructs an LRUCache with the specified capacity.
   * @param {number} capacity - The maximum number of items the cache can hold.
   */
  constructor(capacity: number) {
    this.capacity = capacity;
    this.map = new Map();
    this.head = new CacheNode(0, 0);
    this.tail = new CacheNode(0, 0);
    this.head.next = this.tail;
    this.tail.prev = this.head;
  }

  /**
   * Adds a node to the end of the doubly linked list.
   * @param {CacheNode} node - The node to be added.
   */
  add(node: CacheNode): void {
    const previousEnd = this.tail.prev!;
    previousEnd.next = node;
    node.prev = previousEnd;
    node.next = this.tail;
    this.tail.prev = node;
  }

  /**
   * Removes a node from the doubly linked list.
   * @param {CacheNode} node - The node to be removed.
   */
  remove(node: CacheNode): void {
    node.prev!.next = node.next;
    node.next!.prev = node.prev;
  }

  /**
   * Retrieves the value associated with the specified key.
   * If the key exists, the corresponding node is moved to the end of the doubly linked list.
   * @param {number} key - The key whose associated value is to be returned.
   * @returns {number} The value associated with the key, or -1 if the key does not exist.
   */
  get(key: number): number {
    if (!this.map.has(key)) {
      return -1;
    }

    const node = this.map.get(key)!;
    this.remove(node);
    this.add(node);
    return node.val;
  }

  /**
   * Adds a key-value pair to the cache.
   * If the key already exists, the corresponding node is updated and moved to the end of the doubly linked list.
   * If adding the new key-value pair exceeds the cache's capacity, the least recently used item is removed.
   * @param {number} key - The key with which the specified value is to be associated.
   * @param {number} value - The value to be associated with the specified key.
   */
  put(key: number, value: number): void {
    if (this.map.has(key)) {
      const oldNode = this.map.get(key)!;
      this.remove(oldNode);
    }

    const node = new CacheNode(key, value);
    this.map.set(key, node);
    this.add(node);

    if (this.map.size > this.capacity) {
      const nodeToDelete = this.head.next!;
      this.remove(nodeToDelete);
      this.map.delete(nodeToDelete.key);
    }
  }
}
